include/aikido/util/detail/yaml_extension.hpp

```diff
@@ -0,0 +1,248 @@
+#ifndef AIKIDO_UTIL_DETAIL_YAMLEIGENEXTENSION_HPP_
+#define AIKIDO_UTIL_DETAIL_YAMLEIGENEXTENSION_HPP_
+
+#include <sstream>
+#include <unordered_map>
+#include <Eigen/Dense>
+#include <dart/common/StlHelpers.hpp>
+#include <yaml-cpp/yaml.h>
+
+namespace aikido {
+namespace util {
+namespace detail {
+
+//==============================================================================
+template <typename MatrixType, bool IsVectorAtCompileTime>
+struct encode_impl
+{
+  // Nothing defined. This class should be always specialized.
+
+  // The reason this exists is so we can use template specialization to switch
+  // between serializing vectors and serializing matrices based on the
+  // IsVectorAtCompileTime flag. This is a "nice to have" that lets us generate
+  // !Vector [1, 2, 3] instead of the syntactic travesty like
+  // !Matrix [[1], [2] ,[3]] when serializing an Eigen vector.
+};
+
+//==============================================================================
+// Specialization for vector type
+template <typename MatrixType>
+struct encode_impl<MatrixType, true>
+{
+  static YAML::Node encode(const MatrixType& matrix)
+  {
+    using Index = typename MatrixType::Index;
+
+    YAML::Node node;
+    node.SetTag("!Vector");
+
+    for (Index i = 0; i < matrix.size(); ++i)
+      node.push_back(YAML::Node(matrix[i]));
+
+    return node;
+  }
+};
+
+//==============================================================================
+// Specialization for matrix type
+template <typename MatrixType>
+struct encode_impl<MatrixType, false>
+{
+  static YAML::Node encode(const MatrixType& matrix)
+  {
+    using Index = typename MatrixType::Index;
+
+    YAML::Node node;
+    node.SetTag("!Matrix");
+
+    for (Index r = 0; r < matrix.rows(); ++r)
+    {
+      YAML::Node row(YAML::NodeType::Sequence);
+
+      for (Index c = 0; c < matrix.cols(); ++c)
+        row.push_back(matrix(r, c));
+
+      node.push_back(row);
+    }
+
+    return node;
+  }
+};
+
+inline YAML::Mark getMark(const YAML::Node& node)
+{
+#ifdef YAMLCPP_NODE_HAS_MARK
+  return node.Mark();
+#else
+  DART_UNUSED(node);
+  return YAML::Mark::null_mark();
+#endif
+}
+
+} // namespace detail
+} // namespace util
+} // namespace aikido
+
+namespace YAML {
+
+//==============================================================================
+// Specialization for Eigen::Matrix<...>. This enables to use YAML::Node
+// with std::unordered_map as `Node(Eigen::Matrix3d::Identity())` and
+// `node.as<Eigen::Matrix3d>()`.
+template <typename _Scalar,
+          int _Rows,
+          int _Cols,
+          int _Options,
+          int _MaxRows,
+          int _MaxCols>
```

Comment:
@jslee02 Did you determine whether if or not we can replace this with `template <class _Derived> struct<Eigen::MatrixBase<_Derived>>`? I really liked that idea.